streaming/base/dataset/streaming.py

```diff
@@ -0,0 +1,639 @@
+        """
+        shard, index_in_shard = self.index.find_sample(index)
+        reader = self.shards[shard]
+        return reader[index_in_shard]
+
+    def _resume(self, epoch: int) -> Tuple[int, List[NDArray[np.int64]]]:
+        """Resume from checkpoint.
+
+        Args:
+            epoch (int): What epoch we think it is (pre-checkpoint).
+
+        Returns:
+            Tuple[int, List[NDArray[np.int64]]]: Pair of (resumed epoch, old sessions).
+        """
+        world = World()
+
+        # Get the resume state, if it exists.
+        name = f'{self._prefix}_resume'
+        try:
+            shm = SharedMemory(name)
+        except:
+            # There is nothing to resume.
+            return epoch, []
+
+        # Parse the existent resume state.
+        obj = json.loads(bytes(shm.buf).decode('utf-8'))
+
+        # Check if the resume state is stale.
+        if obj['epoch'] < epoch:
+            # Clean up stale state.
+            if world.is_local_leader:
+                shm.unlink()
+            return epoch, []
+
+        # Load the correct epoch and previous training sessions this epoch.
+        epoch = obj['epoch']
+        old_sessions = [np.array(x) for x in obj['sessions']]
+        return epoch, old_sessions
+
+    def _create_cur_session(self, epoch: int) -> Tuple[NDArray[np.int64], SharedMemory]:
+        """Create the current session, as we have just started an epoch.
+
+        Called by __iter__() in a worker process, or a per-rank process if workers aren't used.
+
+        Note: this also returns the underlying shared memory object, because the returned array
+        will become invalidated when it goes out of scope.
+
+        Args:
+            epoch (int): The current epoch.
+
+        Returns:
+            Tuple[NDArray[np.int64], SharedMemory]: Session and handle to shared memory.
+        """
+        world = World()
+        shm_name = f'{self._prefix}_session_{epoch}'
+        shm_bytes = world.num_workers * np.int64().nbytes
+        try:
+            shm = SharedMemory(shm_name, True, shm_bytes)
+        except FileExistsError:
+            shm = SharedMemory(shm_name)
+        cur_session = np.ndarray(world.num_workers, buffer=shm.buf, dtype=np.int64)
+        cur_session[:] = 0
+        return cur_session, shm
+
+    def _lookup_cur_session(
+            self, epoch: int) -> Tuple[Optional[NDArray[np.int64]], Optional[SharedMemory]]:
+        """Look up the current session, which exists if we are currently training.
+
+        Called by state_dict() in a per-rank process.
+
+        Note: this also returns the underlying shared memory object (if it exists), because the
+        returned array will become invalidated when it goes out of scope.
+
+        Args:
+            epoch (int): The current epoch.
+
+        Returns:
+            Tuple[Optional[NDArray[np.int64]], Optional[SharedMemory]]: Maybe session, maybe shm.
+        """
+        shm_name = f'{self._prefix}_session_{epoch}'
```

Comment:
We can move all the naming line to a `getter()` method ? This way it would be easy to understand and modify at later point since it would be at just one place, i.e. in a `getter()` method